### ISSUE TITLE
Using select shortcut and lateral joins with many-to-many relation

### ISSUE THREAD

**Issue by johannessjoberg** [jawj/zapatos#48]
I'm wondering if there is a recommended way of using select shortcuts and lateral joins for many-to-many relations.

In the documentation, I've only found examples of relations that have the foreign key in their table and can therefore utilize the whereable. For example:
```
const authorsBooksTags = await db.select('authors', db.all, {
  lateral: {
    books: db.select('books', { authorId: db.parent('id') }, {
      lateral: {
        tags: db.select('tags', { bookId: db.parent('id') }, { columns: ['tag'] })
      }
    })
  }
}).run(pool);
```

Is there a way to use joins+where to support many-to-many relations?

**Comment by jawj** [comment]
Unless I've missed something, you do very much the same sort of thing you quote above, using nested `lateral` queries.

Take these three tables:

```sql
CREATE TABLE "products" 
( "id" int PRIMARY KEY GENERATED BY DEFAULT AS IDENTITY
, "productName" text NOT NULL
);
CREATE TABLE "orders" 
( "id" int PRIMARY KEY GENERATED BY DEFAULT AS IDENTITY
, "userEmail" text NOT NULL
);
CREATE TABLE "orderProducts" 
( "orderId" int NOT NULL REFERENCES "orders"("id")
, "productId" int NOT NULL REFERENCES "products"("id") 
);
```

Find an order with its products like so:

```typescript
const order = await db.selectOne('orders', { id: savedOrder.id }, {
  lateral: {
    orderProducts: db.select('orderProducts', { orderId: db.parent('id') }, {
      lateral: {
        product: db.selectExactlyOne('products', { id: db.parent('productId') })
      }
    })
  }
}).run(pool);

if (order) console.log(`For ${order.userEmail}:\n${order.orderProducts.map(op => `- ${op.product.productName}`).join('\n')}`);
```
Which prints:
```
For stephen@privacy.net:
- Flushed grollings
- Jigged olive-spantles
- Embarrassed parping couplets
```

Is this what you're looking for?

**Comment by johannessjoberg** [comment]
Thanks! I was hoping to be able to skip the join table property, in this case, `orderProducts`, to avoid re-formatting the response with code. Is there a way to get this structure from the select straight away (based on your example):
```
// This
{
  id: <orderId>,
  products: [{ id: <productId> }]
}

// Not this
{
  id: <orderId>,
  orderProducts: [
    { product: { id: <productId> } }
  ]
}
```

**Comment by jawj** [comment]
OK, no, there isn't currently a way to do that. We'd need to find an appropriate SQL syntax and type inference for that case. I'll have a think about it.

**Comment by jawj** [comment]
OK, this turns out to be fairly simple to support.

I've had a play with making the `lateral` option also accept, as an alternative to the mapping of names to subqueries it already accepts, a single direct subquery. In the direct subquery case, the result of the subquery is passed through and overwrites the result of the containing query.

This is probably clearer in action:

```typescript
const order = await db.selectOne('orders', { id: savedOrder.id }, {
  lateral: {
    products: db.select('orderProducts', { orderId: db.parent('id') }, {
      lateral: db.selectExactlyOne('products', { id: db.parent('productId') })  // direct subquery
    })
  }
}).run(pool);
```

Which gives this result (correctly typed):

```
{
  id: 14,
  products: [
    { id: 43, productName: 'Flushed grollings' },
    { id: 44, productName: 'Jigged olive-spantles' },
    { id: 45, productName: 'Embarrassed parping couplets' }
  ],
  userEmail: 'stephen@privacy.net'
}
```

You can try this by checking out using `npm link` on the branch `many-to-many-experiment`.

My reservation is that I'm not sure this behaviour is hugely intuitive. I wonder about using a new option (`passthru`?) to accept the single direct subquery case, although that may complicate the typings a bit.

**Comment by jawj** [comment]
Actually, I tell a lie, it's not correctly typed yet: `order.products` is currently typed `any` ...

**Comment by johannessjoberg** [comment]
Nice! I think the direct subquery looks good. 

A similar approach to a new `passthru` option could be to introduce something like `lateralThru`:
```
const order = await db.selectOne('orders', { id: savedOrder.id }, {
  lateralThru: {
    thru: db.select('orderProducts', { orderId: db.parent('id') }),
    select: db.selectExactlyOne('products', { id: db.parent('productId') })
  }
}).run(pool);
```

**Comment by jawj** [comment]
OK, I've stuck with the original design I showed above, which has grown on me a little, and this is now released in v3.1 ([documentation](https://jawj.github.io/zapatos/#lateral-pass-through)).

**Comment by johannessjoberg** [comment]
It looks great. Thank you!
